### ISSUE TITLE
NullReferenceException when changing Position from ViewModel just before page is popped from the stack

### ISSUE THREAD

**Issue by bart-256** [alexrainman/CarouselView#404]
Hi,
So I am experiencing this issue on iOS when page is about to be popped from the stack and I'm altering the Position property from my ViewModel just before it happens then I'm getting following exception:

```
Unhandled Exception:
System.NullReferenceException: Object reference not set to an instance of an object
  at CarouselView.FormsPlugin.iOS.CarouselViewRenderer.Scroller_Scrolled (System.Object sender, System.EventArgs e) [0x00012] in <ce79f845da2443459a9bcb2bc159a217>:0 
  at UIKit.UIScrollView+_UIScrollViewDelegate.Scrolled (UIKit.UIScrollView scrollView) [0x0000a] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.10.1.178/src/Xamarin.iOS/UIKit/UIScrollView.g.cs:1360 
--- End of stack trace from previous location where exception was thrown ---
  at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.10.1.178/src/Xamarin.iOS/UIKit/UIApplication.cs:79 
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.10.1.178/src/Xamarin.iOS/UIKit/UIApplication.cs:63 
  at MyProject.iOS.Application.Main (System.String[] args) [0x00001] in /Users/mac/Projects/myproject/MyProject/MyProject.iOS/Main.cs:12 
2018-05-22 14:43:28.225 MyProject.iOS[82399:2342591] Unhandled managed exception:
Object reference not set to an instance of an object (System.NullReferenceException)
  at CarouselView.FormsPlugin.iOS.CarouselViewRenderer.Scroller_Scrolled (System.Object sender, System.EventArgs e) [0x00012] in <ce79f845da2443459a9bcb2bc159a217>:0 
  at UIKit.UIScrollView+_UIScrollViewDelegate.Scrolled (UIKit.UIScrollView scrollView) [0x0000a] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.10.1.178/src/Xamarin.iOS/UIKit/UIScrollView.g.cs:1360 
--- End of stack trace from previous location where exception was thrown ---
  at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.10.1.178/src/Xamarin.iOS/UIKit/UIApplication.cs:79 
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.10.1.178/src/Xamarin.iOS/UIKit/UIApplication.cs:63 
  at MyProject.iOS.Application.Main (System.String[] args) [0x00001] in /Users/mac/Projects/myproject/MyProject/MyProject.iOS/Main.cs:12 
```

I've implemented simple slideshow in my ViewModel:

```
        private void RunSlideshow()
        {
            if (_cts != null)
            {
                return;
            }

            _cts = new CancellationTokenSource();

            var token = _cts.Token;

            _slideshowTask = Task.Run(async () =>
            {
                while (!token.IsCancellationRequested)
                {
                    await Task.Delay(2000); //2s

                    if (token.IsCancellationRequested)
                    {
                        return;
                    }

                    await ThreadHelper.RunOnMainThreadAsync(() =>
                    {
                        if (token.IsCancellationRequested)
                        {
                            return Task.CompletedTask;
                        }

                        try
                        {
                            if (Position + 1 >= ImageList.Count)
                            {
                                Position = 0;
                            }
                            else
                            {
                                Position = Position + 1;
                            }
                        } catch {}

                        return Task.CompletedTask;
                    });
                }
            }, token);
        }

        private void StopSlideshow()
        {
            _cts?.Cancel();
            _cts?.Dispose();
            _cts = null;
        }

        public Task OnDisappearing()
        {
            StopSlideshow();

            return Task.CompletedTask;
        }
```

So the exception only occurs when I press back button in some particular moment - it doesn't throw exception each time when I navigate back, just occasionally and I believe it is the moment when code executes just before page is popped. I've added StopSlideshow() to OnDisappearing callback which cancels the slideshow task and added checks token.IsCancellationRequested in many places as you can see from my code. Even surrounded Position changing logic with try catch, but it still throws the same unhandled exception (sometimes).

Plugin version: **5.2.0**

**Comment by AndreiMisiukevich** [comment]
@bart-256 hi, try this plugin instead https://github.com/AndreiMisiukevich/CardView

Because, this plugin was merged to XF..

**Comment by PeteVasi** [comment]
We've been unable to reproduce this ourselves, but we do get crash reports through AppCenter that look like this issue too:
```CarouselViewRenderer.Scroller_Scrolled (System.Object sender, System.EventArgs e)
System.NullReferenceException: Object reference not set to an instance of an object

CarouselViewRenderer.Scroller_Scrolled (System.Object sender, System.EventArgs e)
UIScrollView+_UIScrollViewDelegate.Scrolled (UIKit.UIScrollView scrollView)
(wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate)
UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName)
Application.Main (System.String[] args)
```

Following to see if this gets resolved...

**Comment by NiknakSi** [comment]
I have this reproducible in an app I'm developing at the moment, but annoyingly only on a physical iPhone 6 Plus running iOS 11.1.2. It's fine on other devices and in various simulator devices running iOS 11 and 10, and it's fine on Android. For me it's occurring when I rotate the phone while there's a modal being displayed (a video player in my case).

Once I've debugged it I'll submit a PR with a solution.

**Comment by NiknakSi** [comment]
So in my case this was the Element property being null and the exception being thrown in CarouselView.FormsPlugin.iOS.CarouselViewRenderer by

`if (Element.Orientation == CarouselViewOrientation.Horizontal)`

I've added a little safety which has solved the problem for me. Hopefully this also cures it for you guys, if anyone would be able to test it out that might be handy. Thanks!

**Comment by PeteVasi** [comment]
@alexrainman I see that the suggested PR was closed but not merged.  Any other details on a fix?

**Comment by NiknakSi** [comment]
@PeteVasi Judging by this PR it looks like it's in the process of being merged with Xamarin Forms, but it looks like it's taking some considerable time to get there. In the meantime I guess this project has been abandoned? I appreciate the difficulty of keeping them both in parallel but until it's actually released as part of XF I think fixes should still go into this repo as well, since it's all we've got.

https://github.com/xamarin/Xamarin.Forms/pull/1882

@alexrainman It might also help if there was some kind of official statement on this repo that the code is being merged into XF (sorry if I've missed that somewhere). Closing the PR without any comment is a little frustrating.

I've started maintaining my own fork after also adding the ability to control the slide transition speed on Android (see https://github.com/NiknakSi/CarouselView/tree/niknaksi/configurable-transition-speed).

**Comment by alexrainman** [comment]
@PeteVasi @bart-256 expect a fix for this in the next release.

**Comment by alexrainman** [comment]
@PeteVasi it was merged manually.

**Comment by bluekuen** [comment]
@alexrainman Hello Alex, first, thank you for this great plugin ! Can you approximately tell when you will publish the next release with this fix ? I have many crashes in my production app because of this bug and would love to be free of it. I dont'T want to make stress to you, i just would like to know if it is coming out in the next two or three weeks. 
Thank you and Kind Regards
David
